core/comp_ui.py

```diff
@@ -31,17 +30,22 @@
 
 
 def _PromptLen(prompt_str):
-  """Ignore all characters between \x01 and \x02."""
+  """Ignore all characters between \x01 and \x02 and handle unicode characters.
+  In particular, the display width of a string may be different from either the
+  number of bytes or the number of unicode characters."""
   escaped = False
   length = 0
+  chunk = ""
   for c in prompt_str:
     if c == '\x01':
       escaped = True
+      length += libc.wcswidth(chunk)
+      chunk = ""
     elif c == '\x02':
       escaped = False
     elif not escaped:
-      length += 1
-  return length
+      chunk += c
+  return length + libc.wcswidth(chunk)
```

Comment:
It seems like it would be simpler to just collect all the characters first, and make a single wcswidth() call at the end.  That seems like the same thing.

Reply:
It would do the same thing, yeah. My reasoning is that it would use less memory not to copy the prompt string, but if `setlocale` is expensive it probably makes sense to do it all at once.